In [1]:
import nltk
import pandas as pd
# nltk.download()

In [2]:
dir(nltk)

['AbstractLazySequence',
 'AffixTagger',
 'AlignedSent',
 'Alignment',
 'AnnotationTask',
 'ApplicationExpression',
 'Assignment',
 'BigramAssocMeasures',
 'BigramCollocationFinder',
 'BigramTagger',
 'BinaryMaxentFeatureEncoding',
 'BlanklineTokenizer',
 'BllipParser',
 'BottomUpChartParser',
 'BottomUpLeftCornerChartParser',
 'BottomUpProbabilisticChartParser',
 'Boxer',
 'BrillTagger',
 'BrillTaggerTrainer',
 'CFG',
 'CRFTagger',
 'CfgReadingCommand',
 'ChartParser',
 'ChunkParserI',
 'ChunkScore',
 'Cistem',
 'ClassifierBasedPOSTagger',
 'ClassifierBasedTagger',
 'ClassifierI',
 'ConcordanceIndex',
 'ConditionalExponentialClassifier',
 'ConditionalFreqDist',
 'ConditionalProbDist',
 'ConditionalProbDistI',
 'ConfusionMatrix',
 'ContextIndex',
 'ContextTagger',
 'ContingencyMeasures',
 'CoreNLPDependencyParser',
 'CoreNLPParser',
 'Counter',
 'CrossValidationProbDist',
 'DRS',
 'DecisionTreeClassifier',
 'DefaultTagger',
 'DependencyEvaluator',
 'DependencyGrammar',
 'DependencyGrap

O que podemos fazer com o NLTK?

In [3]:
from nltk.corpus import stopwords

stopwords.words('english')

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

## Básico do NLP: Lendo dados de texto e por que devemos limpar o texto?

### Lendo dados de texto semi-estruturados

In [4]:
data = open(".\\ex_files\\Ch01\\01_03\\Start\\SMSSpamCollection.tsv").read()

data[0:500]

"ham\tI've been searching for the right words to thank you for this breather. I promise i wont take your help for granted and will fulfil my promise. You have been wonderful and a blessing at all times.\nspam\tFree entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's\nham\tNah I don't think he goes to usf, he lives around here though\nham\tEven my brother is not like to speak with me. They treat me like aid"

In [5]:
dados_analisados = data.replace('\t', '\n').split('\n')

In [6]:
dados_analisados[0:5]

['ham',
 "I've been searching for the right words to thank you for this breather. I promise i wont take your help for granted and will fulfil my promise. You have been wonderful and a blessing at all times.",
 'spam',
 "Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's",
 'ham']

Isso nos possibilita criar uma lista com uma melhor estrutura.

In [7]:
label_lista = dados_analisados[0::2]
texto_lista = dados_analisados[1::2]

In [8]:
print(label_lista[0:5])
print(texto_lista[0:5])

['ham', 'spam', 'ham', 'ham', 'ham']
["I've been searching for the right words to thank you for this breather. I promise i wont take your help for granted and will fulfil my promise. You have been wonderful and a blessing at all times.", "Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive entry question(std txt rate)T&C's apply 08452810075over18's", "Nah I don't think he goes to usf, he lives around here though", 'Even my brother is not like to speak with me. They treat me like aids patent.', 'I HAVE A DATE ON SUNDAY WITH WILL!!']


Com isso, podemos criar uma base de dados do pandas para organizar os resultados. Mas antes, temos que verificar se as listas possuem o mesmo tamanho, para que possamos criar uma base de dados sem grandes problemas

In [9]:
print(len(label_lista))
print(len(texto_lista))

5571
5570


Podemos observar que a variável *label_lista* possui um elemento a mais que a *texto_lista*. Antes de tomarmos alguma decisão sobre como agir diante dessa discrepância do número de elementos, vamos investigar para o *label_lista* quais os elementos presentes na lista. Isso, pois estamos esperando que este possua apenas dois únicos valores: *ham* e *spam*. 

In [10]:
label_set = set(label_lista)
label_set

{'', 'ham', 'spam'}

Podemos observar que há um elemento nulo, não esperado para os tipos de dados que estamos trabalhando para esta lista. Podemos tentar rastrear quantos elementos nulos possuem na lista, de modo a verificar se há somente um elemento nulo (o que explicaria a contagem do elemento a mais, nos levando a resolver este problema).

In [11]:
label_lista.index('')

5570

Podemos observar que a primeira ocorrência do elemento nulo aparece na posição *5570* da lista, ou seja, comparando com o tamanho da lista, este se apresenta no final da lista, como podemos observar abaixo

In [12]:
label_lista[-5:]

['ham', 'ham', 'ham', 'ham', '']

Podemos, podemos simplesmente dropar o último termo.

In [13]:
label_lista = label_lista[:-1]

In [14]:
label_lista[-5:]

['spam', 'ham', 'ham', 'ham', 'ham']

Agora o tamanho da *label_lista* é igual ao tamanho da *texto_lista*.

In [15]:
print(len(label_lista))
print(len(texto_lista))

5570
5570


In [16]:
full_corpus = pd.DataFrame({'label': label_lista, 'texto': texto_lista})

full_corpus.head()

label                                              texto
0   ham  I've been searching for the right words to tha...
1  spam  Free entry in 2 a wkly comp to win FA Cup fina...
2   ham  Nah I don't think he goes to usf, he lives aro...
3   ham  Even my brother is not like to speak with me. ...
4   ham                I HAVE A DATE ON SUNDAY WITH WILL!!

## Explorando a Base de Dados

In [17]:
# Qual é o formato do dataset?

print("O dado de entrada tem {} linhas e {} colunas".format(full_corpus.shape[0], full_corpus.shape[1]))

O dado de entrada tem 5570 linhas e 2 colunas


In [18]:
# Quantos dados são spam e quantos são ham?

print("De {} dados de entrada, {} são spams e {} são hams".format(len(full_corpus), full_corpus['label'].value_counts()[1], full_corpus['label'].value_counts()[0]))

De 5570 dados de entrada, 746 são spams e 4824 são hams


In [19]:
# Quantos dados faltantes existem?
print("Numero de dados faltantes em label: {}".format(full_corpus['label'].isnull().sum()))
print("Numero de dados faltantes em texto: {}".format(full_corpus['texto'].isnull().sum()))

Numero de dados faltantes em label: 0
Numero de dados faltantes em texto: 0


## Aprendendo como Usar Expressões Regulares

In [20]:
import re

re_teste = "This is a made up string to test 2 different regex methods"
re_test_messy = 'This              is a made up           string to test 2    different regex methods'
re_test_messy_1 = 'This-is-a-made/up.string*to>>>>>>>test-----2"""""different-regex-methods'

In [21]:
# Dividindo a sentença em uma lista de palavras
re.split("\s", re_teste)

['This',
 'is',
 'a',
 'made',
 'up',
 'string',
 'to',
 'test',
 '2',
 'different',
 'regex',
 'methods']

In [22]:
re.split('\s+', re_test_messy)

['This',
 'is',
 'a',
 'made',
 'up',
 'string',
 'to',
 'test',
 '2',
 'different',
 'regex',
 'methods']

In [23]:
re.split('\W+', re_test_messy_1)

['This',
 'is',
 'a',
 'made',
 'up',
 'string',
 'to',
 'test',
 '2',
 'different',
 'regex',
 'methods']

In [24]:
re.findall('\S+', re_teste)

['This',
 'is',
 'a',
 'made',
 'up',
 'string',
 'to',
 'test',
 '2',
 'different',
 'regex',
 'methods']

In [25]:
re.findall("\S+", re_test_messy)

['This',
 'is',
 'a',
 'made',
 'up',
 'string',
 'to',
 'test',
 '2',
 'different',
 'regex',
 'methods']

In [26]:
re.findall("\w+", re_test_messy_1)

['This',
 'is',
 'a',
 'made',
 'up',
 'string',
 'to',
 'test',
 '2',
 'different',
 'regex',
 'methods']

In [27]:
#Substituindo uma string específica
pep8_string = "I try to follow PEP8 guidelines"
pep7_string = "I try to follow PEP7 guidelines"
peep8_string = "I try to follow PEEP8 guidelines"

In [28]:
re.findall('[a-z]+', pep8_string)

['try', 'to', 'follow', 'guidelines']

In [29]:
re.findall('[A-Z]+', pep8_string)

['I', 'PEP']

In [30]:
# letras de A a Z capitalizadas ou números de 0 a 9
re.findall('[A-Z0-9]+', pep8_string)

['I', 'PEP8']

In [31]:
# letras de A a Z capitalizadas e números de 0 a 9
re.findall('[A-Z]+[0-9]+', pep8_string)

['PEP8']

In [32]:
re.sub('[A-Z]+[0-9]+', 'PEP8 Python Styleguide', pep8_string)

'I try to follow PEP8 Python Styleguide guidelines'

In [34]:
re.sub('[A-Z]+[0-9]+', 'PEP8 Python Styleguide', pep8_string)

'I try to follow PEP7 Python Styleguide guidelines'

In [35]:
re.sub('[A-Z]+[0-9]+', 'PEEP8 Python Styleguide', peep8_string)

'I try to follow PEEP8 Python Styleguide guidelines'

## Implementando um Pipeline para Limpar o Texto

### Pré-processamento dos Dados de Texto
Para limpar os dados de texto, é necessário que evidenciemos algumas características que queremos que o nosso modelo de Machine Learning aprenda. Limpando (ou realizando um pré-processamento) nos dados de texto, geralmente, consiste nos seguintes passos:

1. Remover pontuações
2. Tokenização
3. Remover stopwords
4. Lemmatização/Stemização

Os primeiros três passos serão feitos nesta parte, pois eles são implementados na grande maioria de qualquer pipeline de limpeza de texto. Lematização e Stemização serão implementados posteriormente, tendo em vista que eles são úteis, porém não críticos para o bom funcionamento do modelo.

In [46]:
pd.set_option('display.max_colwidth', 100)

data=pd.read_csv(".\\ex_files\\Ch01\\01_03\\Start\\SMSSpamCollection.tsv", sep='\t', header=None)
data.columns = ['label', 'texto']

data.head()

label  \
0   ham   
1  spam   
2   ham   
3   ham   
4   ham   

                                                                                                 texto  
0  I've been searching for the right words to thank you for this breather. I promise i wont take yo...  
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive ...  
2                                        Nah I don't think he goes to usf, he lives around here though  
3                        Even my brother is not like to speak with me. They treat me like aids patent.  
4                                                                  I HAVE A DATE ON SUNDAY WITH WILL!!

In [ ]:
# Como um dado limpo deve parecer?
data_limpo = pd.read_csv(".\\ex_files\\Ch01\\01_09\\Start\\SMSSpamCollection_cleaned.tsv", sep='\t', header=None)
data_limpo.head()

0  \
0  label   
1    ham   
2   spam   
3    ham   
4    ham   

                                                                                                     1  \
0                                                                                            body_text   
1  I've been searching for the right words to thank you for this breather. I promise i wont take yo...   
2  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive ...   
3                                        Nah I don't think he goes to usf, he lives around here though   
4                        Even my brother is not like to speak with me. They treat me like aids patent.   

                                                                                                     2  
0                                                                                     body_text_nostop  
1  ['ive', 'searching', 'right', 'words', 'thank', 'breather', 'promise', 'wont', 'take', 'help', '...  
2  ['free', 'entry', '2', 'wkly', 'comp', 'win', 'fa', 'cup', 'final', 'tkts', '21st', 'may', '2005...  
3                                 ['nah', 'dont', 'think', 'goes', 'usf', 'lives', 'around', 'though']  
4                              ['even', 'brother', 'like', 'speak', 'treat', 'like', 'aids', 'patent']

### Removendo a pontuação

In [39]:
import string
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [47]:
def remove_punct(text):
  text_nopunct = "".join([char for char in text if char not in string.punctuation])
  return text_nopunct

data['texto_limpo'] = data['texto'].apply(lambda x: remove_punct(x))
data

label  \
0      ham   
1     spam   
2      ham   
3      ham   
4      ham   
...    ...   
5563  spam   
5564   ham   
5565   ham   
5566   ham   
5567   ham   

                                                                                                    texto  \
0     I've been searching for the right words to thank you for this breather. I promise i wont take yo...   
1     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive ...   
2                                           Nah I don't think he goes to usf, he lives around here though   
3                           Even my brother is not like to speak with me. They treat me like aids patent.   
4                                                                     I HAVE A DATE ON SUNDAY WITH WILL!!   
...                                                                                                   ...   
5563  This is the 2nd time we have tried 2 contact u. U have won the £750 Pound prize. 2 claim is easy...   
5564                                                                 Will ü b going to esplanade fr home?   
5565                                            Pity, * was in mood for that. So...any other suggestions?   
5566  The guy did some bitching but I acted like i'd be interested in buying something else next week ...   
5567                                                                           Rofl. Its true to its name   

                                                                                              texto_limpo  
0     Ive been searching for the right words to thank you for this breather I promise i wont take your...  
1     Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive e...  
2                                             Nah I dont think he goes to usf he lives around here though  
3                             Even my brother is not like to speak with me They treat me like aids patent  
4                                                                       I HAVE A DATE ON SUNDAY WITH WILL  
...                                                                                                   ...  
5563  This is the 2nd time we have tried 2 contact u U have won the £750 Pound prize 2 claim is easy c...  
5564                                                                  Will ü b going to esplanade fr home  
5565                                                   Pity  was in mood for that Soany other suggestions  
5566  The guy did some bitching but I acted like id be interested in buying something else next week a...  
5567                                                                            Rofl Its true to its name  

[5568 rows x 3 columns]

### Tokenização

In [48]:
def tokenize(text):
  tokens = re.split('\W+', text)
  return tokens


data['texto_tokenizado'] = data['texto_limpo'].apply(lambda x: tokenize(x.lower()))

In [49]:
data.head()

label  \
0   ham   
1  spam   
2   ham   
3   ham   
4   ham   

                                                                                                 texto  \
0  I've been searching for the right words to thank you for this breather. I promise i wont take yo...   
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive ...   
2                                        Nah I don't think he goes to usf, he lives around here though   
3                        Even my brother is not like to speak with me. They treat me like aids patent.   
4                                                                  I HAVE A DATE ON SUNDAY WITH WILL!!   

                                                                                           texto_limpo  \
0  Ive been searching for the right words to thank you for this breather I promise i wont take your...   
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive e...   
2                                          Nah I dont think he goes to usf he lives around here though   
3                          Even my brother is not like to speak with me They treat me like aids patent   
4                                                                    I HAVE A DATE ON SUNDAY WITH WILL   

                                                                                      texto_tokenizado  
0  [ive, been, searching, for, the, right, words, to, thank, you, for, this, breather, i, promise, ...  
1  [free, entry, in, 2, a, wkly, comp, to, win, fa, cup, final, tkts, 21st, may, 2005, text, fa, to...  
2                            [nah, i, dont, think, he, goes, to, usf, he, lives, around, here, though]  
3         [even, my, brother, is, not, like, to, speak, with, me, they, treat, me, like, aids, patent]  
4                                                           [i, have, a, date, on, sunday, with, will]

### Removendo as Stopwords

In [51]:
stopword = nltk.corpus.stopwords.words('english')

In [53]:
def remove_stopwords(tokenized_list):
  text = [word for word in tokenized_list if word not in stopword]
  return text


data['texto_sem_stopwords'] = data['texto_tokenizado'].apply(lambda x: remove_stopwords(x))
data.head()

label  \
0   ham   
1  spam   
2   ham   
3   ham   
4   ham   

                                                                                                 texto  \
0  I've been searching for the right words to thank you for this breather. I promise i wont take yo...   
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive ...   
2                                        Nah I don't think he goes to usf, he lives around here though   
3                        Even my brother is not like to speak with me. They treat me like aids patent.   
4                                                                  I HAVE A DATE ON SUNDAY WITH WILL!!   

                                                                                           texto_limpo  \
0  Ive been searching for the right words to thank you for this breather I promise i wont take your...   
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive e...   
2                                          Nah I dont think he goes to usf he lives around here though   
3                          Even my brother is not like to speak with me They treat me like aids patent   
4                                                                    I HAVE A DATE ON SUNDAY WITH WILL   

                                                                                      texto_tokenizado  \
0  [ive, been, searching, for, the, right, words, to, thank, you, for, this, breather, i, promise, ...   
1  [free, entry, in, 2, a, wkly, comp, to, win, fa, cup, final, tkts, 21st, may, 2005, text, fa, to...   
2                            [nah, i, dont, think, he, goes, to, usf, he, lives, around, here, though]   
3         [even, my, brother, is, not, like, to, speak, with, me, they, treat, me, like, aids, patent]   
4                                                           [i, have, a, date, on, sunday, with, will]   

                                                                                   texto_sem_stopwords  
0  [ive, searching, right, words, thank, breather, promise, wont, take, help, granted, fulfil, prom...  
1  [free, entry, 2, wkly, comp, win, fa, cup, final, tkts, 21st, may, 2005, text, fa, 87121, receiv...  
2                                                 [nah, dont, think, goes, usf, lives, around, though]  
3                                              [even, brother, like, speak, treat, like, aids, patent]  
4                                                                                       [date, sunday]